common/src/main/java/org/apache/celeborn/common/network/server/BufferStreamManager.java

```diff
@@ -313,74 +321,125 @@ public synchronized void setupDataPartitionReader(
-      buffer.clear();
-      bufferQueue.add(buffer);
-      triggerRead();
+    public void recycle(ByteBuf buffer) {
+      bufferQueue.recycle(buffer);
+      if (isReleased || readers.isEmpty() || bufferQueue.isReleased()) {
+        return;
+      }
+
+      if (bufferQueue.size() > 0) {
+        triggerRead();
+      }
+
+      applyNewBuffers();
+    }
+
+    private void applyNewBuffers() {
+      logger.debug(
+          "try to apply new buffers {} {} {} {} {}",
+          isWaitingResources.get(),
+          bufferQueue.numBuffersOccupied(),
+          bufferQueue.size(),
+          readers.size(),
+          bufferQueue.numBuffersOccupied() + minReadBuffers);
+      if (bufferQueue.size() < minReadBuffers
+          && !readers.isEmpty()
+          && bufferQueue.numBuffersOccupied() + minReadBuffers <= maxReadBuffers
+          && isWaitingResources.compareAndSet(false, true)) {
+        logger.debug(
+            "apply new buffers while current buffer queue size {} with read count {} for "
+                + "map data partition {} with active stream id count {}",
+            bufferQueue.size(),
+            readers.size(),
+            this,
+            activeStreamIds.size());
+
+        memoryManager.requestReadBuffers(
+            new ReadBufferRequest(
+                minReadBuffers,
+                maxReadBuffers,
+                fileInfo.getBufferSize(),
+                (allocatedBuffers, throwable) -> onBuffer(allocatedBuffers)));
+      }
     }
 
     public synchronized void readBuffers() {
       if (isReleased) {
-        // some read executor task may already be submitted to the threadpool
+        // some read executor task may already be submitted to the thread pool
         return;
       }
 
       try {
-        PriorityQueue<DataPartitionReader> sortedReaders = new PriorityQueue<>(readers);
-        for (DataPartitionReader reader : readers) {
+        // make sure that all reader are open
+        PriorityQueue<DataPartitionReader> sortedReaders = new PriorityQueue<>(readers.values());
+        for (DataPartitionReader reader : readers.values()) {
           reader.open(dataFileChanel, indexChannel);
         }
-        while (buffers != null && buffers.size() > 0 && !sortedReaders.isEmpty()) {
-          BufferRecycler bufferRecycler =
-              new BufferRecycler(memoryManager, (buffer) -> this.recycle(buffer, buffers));
+        while (bufferQueue.size() > 0 && !sortedReaders.isEmpty()) {
+          BufferRecycler bufferRecycler = new BufferRecycler(MapDataPartition.this::recycle);
           DataPartitionReader reader = sortedReaders.poll();
           try {
-            if (!reader.readAndSend(buffers, bufferRecycler)) {
-              readers.remove(reader);
+            if (!reader.readAndSend(bufferQueue, bufferRecycler)) {
+              // this reader has finished, recycle stream id
+              if (reader.isFinished()) {
```

Comment:
we can remove this as reader finished will close the reader self.

Reply:
We cannot remove this because when readAndSend returns false do not mean that the data has been sent. It only means that data is in memory.